drivers,subsystems: Use kernel heap for allocations.

iabdalkader opened the pull request:
Some native Zephyr subsystems and drivers use malloc instead of the kernel heap, as they should. This is not only inconsistent with other subsystems and drivers, but also problematic because:

a) It may introduce a security risk, since libc heap memory is user-accessible
b) Those allocations can fail if `CONFIG_COMMON_LIBC_MALLOC_ARENA_SIZE=0`.

Fixes #100112

* I wasn't sure about native simulators, but I fixed them anyway.

* I'm not sure about the following files. Are these considered kernel or user-space?
  * lib/posix/options/env_common.c
  * lib/posix/options/net.c
  * subsys/fs/fuse_fs_access_bottom.c

* Finally, some of the files use `strdup`, which returns malloc'd memory. For example, `drivers/net/nsos_adapt.c` and `subsys/net/lib/ocpp/key_mgmt.c`. Fixing those is more tricky (would need to k_malloc, check for NULL, strncpy etc..). I've also noticed the memory returned from `strdup` is never checked for NULL, perhaps the variables to which this memory is assigned can be NULL. It's hard to say, we need someone more familiar with the context of those files.

iabdalkader committed:
drivers: wifi: Use kernel heap for allocations

Use the kernel heap instead of the libc heap, improving
security and consistency.

Signed-off-by: Ibrahim Abdalkader <i.abdalkader@gmail.com>

iabdalkader committed:
mgmt: mcumgr: grp: Use kernel heap for allocations

Use the kernel heap instead of the libc heap, improving
security and consistency.

Signed-off-by: Ibrahim Abdalkader <i.abdalkader@gmail.com>

iabdalkader committed:
sensing: sensor_mgmt: Use kernel heap for allocations

Use the kernel heap instead of the libc heap, improving
security and consistency.

Signed-off-by: Ibrahim Abdalkader <i.abdalkader@gmail.com>

iabdalkader commented on drivers/eeprom/eeprom_simulator_native.c:
```diff
@@ -22,6 +22,7 @@
 #include <stdlib.h>
 #include <stdbool.h>
 #include <string.h>
+#include <zephyr/kernel.h>
```
Thanks! Will revert those changes, and fix the failing unit-tests/samples. Any comment on `strdup` usage?

jukkar commented:
> * Finally, some of the files use `strdup`, which returns malloc'd memory. For example, `drivers/net/nsos_adapt.c` and `subsys/net/lib/ocpp/key_mgmt.c`. Fixing those is more tricky (would need to k_malloc, check for NULL, strncpy etc..). I've also noticed the memory returned from `strdup` is never checked for NULL, perhaps the variables to which this memory is assigned can be NULL. It's hard to say, we need someone more familiar with the context of those files.

About the `strdup()` usage, as you said sometimes it is ok and the code can handle NULL string so it might not be an issue here. For the OCPP lib, @ssekar15 could you check if code will be ok if the `strdup` calls return NULL.

JarmouniA reviewed (commented):
The enabling of the kernel heap for a given sample/test should happen in the same commit where its underlying subsys/driver is modified to use k_heap.

XenuIsWatching commented on drivers/i3c/i3c_common.c:
```diff
@@ -1295,7 +1296,7 @@ int i3c_bus_deftgts(const struct device *dev)
 
 	/* Allocate memory for the struct with enough space for the targets */
-	deftgts = malloc(data_len);
+	deftgts = k_malloc(data_len);
```
I agree with these changes... but my concern with this is that by default `CONFIG_HEAP_MEM_POOL_SIZE` is 0 and this will fail everytime, this probably should then be doing an adding of some bytes as followed by here (https://docs.zephyrproject.org/latest/kernel/memory_management/heap.html#defining-the-heap-memory-pool)

Maybe 32 bytes for default for I3C

iabdalkader replied:
I don't see that option actually being used anywhere, and from the description alone I don't quite understand what it should do 😅 

https://github.com/search?q=repo%3Azephyrproject-rtos%2Fzephyr%20HEAP_MEM_POOL_ADD_SIZE_UHC_MCUX&type=code

Does it get used by linker scripts somehow? If you let me know what needs to be done, I'll update the PR no problem.

nordicjm replied:
They are collated and used yes https://github.com/zephyrproject-rtos/zephyr/blob/main/kernel/CMakeLists.txt#L135

iabdalkader replied:
Never mind, I see plenty of `HEAP_MEM_POOL_ADD_SIZE_*`, I'll figure it out.

iabdalkader commented on subsys/mgmt/mcumgr/grp/settings_mgmt/src/settings_mgmt.c:
```diff
@@ -75,12 +75,12 @@ static int settings_mgmt_read(struct smp_streamer *ctxt)
-	key_name = (char *)malloc(key.len + 1);
-	data = (uint8_t *)malloc(max_size);
+	key_name = (char *)k_malloc(key.len + 1);
+	data = (uint8_t *)k_malloc(max_size);
```
Is it specifically libc heap or more like any heap vs the stack? In this case, maybe the kconfig option's `depend`, and its description, should be fixed.

nordicjm replied:
that would be fine too (as per edited comment)

iabdalkader committed:
drivers: i3c: Use kernel heap for allocations

Use the kernel heap instead of the libc heap, improving
security and consistency.

Signed-off-by: Ibrahim Abdalkader <i.abdalkader@gmail.com>

XenuIsWatching reviewed (dismissed):
Approving on the I3C changes here. I can't really comment on the other areas here

nordicjm reviewed (dismissed):
MCUmgr changes OK

JarmouniA commented on drivers/i3c/Kconfig:
```diff
@@ -58,6 +58,13 @@ config I3C_TARGET_BUFFER_MODE
+	default 48
+	depends on I3C_CONTROLLER && I3C_TARGET
+	help
+	  The size of the memory pool used by I3C. Used only for the ccc deftgts.
```
"ccc deftgts"!?

JarmouniA commented on drivers/i3c/Kconfig:
```diff
@@ -58,6 +58,13 @@ config I3C_TARGET_BUFFER_MODE
+config HEAP_MEM_POOL_ADD_SIZE_I3C
+	int "Heap memory pool size for I3C"
+	default 48
+	depends on I3C_CONTROLLER && I3C_TARGET
```
```suggestion
	depends on I3C_CONTROLLER || I3C_TARGET
```
!?

dcpleung replied:
This really needs to be *OR*.

iabdalkader replied:
Maybe it supports some dual role mode?

dcpleung replied:
Hm... I think it should only be enabled for `CONFIG_I3C_CONTROLLER` as `DEFTGTS` are only for controllers. It is for the active controller to communicate how many targets are on the bus to the secondary controller.

jukkar reviewed (dismissed):
ack to wifi changes

iabdalkader commented on drivers/i3c/Kconfig:
```diff
@@ -58,6 +58,13 @@ config I3C_TARGET_BUFFER_MODE
+	default 48
+	depends on I3C_CONTROLLER && I3C_TARGET
+	help
+	  The size of the memory pool used by I3C. Used only for the ccc deftgts.
```
Can I fix this in a followup? It was hard enough getting that many approvals...

JarmouniA replied:
Fine by me.

iabdalkader replied:
I fixed this now, since I had to rebase anyway.

yperess reviewed (dismissed):
Lgtm from a sensor perspective

dcpleung commented:
Need to rebase due to conflicts.

lixuzha reviewed (approved):
Lgtm from a sensing subsys perspective.
